File: 3122. Minimum Number of Operations to Satisfy Conditions.go
func minimumOperations(grid [][]int) int {
	n, m := len(grid), len(grid[0])
	cost := make([][]int, m)
	for j := range cost {
		cost[j] = make([]int, 10)
		freq := make([]int, 10)
		for i := range n {
			freq[grid[i][j]] += 1
		}
		for i := range 10 {
			cost[j][i] = n - freq[i]
		}
	}
	cache := make(map[[2]int]int)
	return helper(cost, 0, -1, cache)
}
func helper(cost [][]int, current, previous int, cache map[[2]int]int) int {
	if current == len(cost) {
		return 0
	}
	if val, ok := cache[[2]int{current, previous}]; ok {
		return val
	}
	answer := math.MaxInt32
	for i := range 10 {
		if i == previous {
			continue
		}
		answer = min(answer, cost[current][i]+helper(cost, current+1, i, cache))
	}
	cache[[2]int{current, previous}] = answer
	return answer
}

File: 3122. Minimum Number of Operations to Satisfy Conditions.java
class Solution {
    public int minimumOperations(int[][] grid) {
        int n = grid.length, m = grid[0].length;
        int[][] cost = new int[m][10];
        for (int j = 0; j < m; j++) {
            int[] freq = new int[10];
            for (int i = 0; i < n; i++) {
                freq[grid[i][j]]++;
            }
            for (int i = 0; i < 10; i++) {
                cost[j][i] = n-freq[i];
            }
        }
        Integer[][] cache = new Integer[m][11];
        return helper(cost, 0, 10, cache);
    }
    private int helper(int[][] cost, int current, int previous, Integer[][] cache) {
        if (current >= cost.length) return 0;
        if (cache[current][previous] != null) return cache[current][previous];
        int answer = Integer.MAX_VALUE;
        for (int i = 0; i < 10; i++) {
            if (i == previous) continue;
            answer = Math.min(answer, cost[current][i] + helper(cost, current + 1, i, cache));
        }
        return cache[current][previous] = answer;
    }
}